[AST] Preserve sugar for generic typealiases.

dcci opened the pull request:
<rdar://problem/43110802>

Commit:
Propagate concrete types for all arguments of an apply instruction

Commit:
Propagate concrete types for all arguments of an apply instruction

benlangmuir committed:
[unittest] Fix some unintentional behaviour in EditingTest

I haven't found any cases where this changes the visible behaviour of
the test, but having the file remain open across invocations was
certainly unexpected and affects how the first part of the test works at
runtime.

benlangmuir committed:
[sourcekit] Fix a race between AST processing and reopening a document

When a document is opened twice without closing it we re-initialize its
internal data. Make sure we don't race with code that accesses that data.

Found by TSan!

jckarter committed:
SILGen: Emit convenience initializers as allocating entry points.

And only dispatch designated inits by their allocating entry points. rdar://problem/29634243

palimondo committed:
[benchmark] Gardening: Code format class Timer

palimondo committed:
[benchmark] Gardening: Extract constant oneSecond

palimondo committed:
[benchmark] Gardening: Timer Parasite Control

Improve conformance to Swift Naming Guidelines
https://swift.org/documentation/api-design-guidelines/#parameter-names

Removed the gross undescores and ticks from parameter names. Ticks are ectoparasites feeding on the blood. We are just measuring time and there is also no [mysterious ticking noise](http://bit.ly/TickNoise) here either…

palimondo committed:
[benchmark] Gardening: Indentation of .listTests

palimondo committed:
[benchmark] Refactor mean and stdev computation

palimondo committed:
[benchmark] Fix index computation for quantiles

Turns out that both the old code in `DriverUtils` that computed median, as well as newer quartiles in `PerformanceTestSamples` had off-by-1 error.

It trully is the 3rd of the 2 hard things in computer science!

palimondo committed:
[benchmark] Refactor min max median computation

We can spare 2 array passes (for min and max), if we just sort first.

palimondo committed:
[benchmark] Move stats computation to BenchResults

palimondo committed:
[benchmark] Refactor to currency type Int

Removed unnecessary use of UInt64, where appropriate, following the advice from Swift Language Guide:

> Use the `Int` type for all general-purpose integer constants and variables in your code, even if they’re known to be nonnegative. Using the default integer type in everyday situations means that integer constants and variables are immediately interoperable in your code and will match the inferred type for integer literal values.
https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html#ID324

palimondo committed:
[benchmark] Gardening: DRYer verbose log

palimondo committed:
[benchmark] Gardening: afterRunSleep is UInt32

palimondo committed:
[benchmark] Gardening: numSamples UInt vs Int

Type check command line argument to be non-negative, but store value in currency type `Int`.

palimondo committed:
[benchmark] Gardening: scale was always Int

Since the `scale` (or `numIters`) is passed to the `test.runFunction` as `Int`, the whole type-casting dance here was just silly!

palimondo committed:
[benchmark] Refactor to report samples in μs

Moved the adjustment of `lastSampleTime` to account for the `scale` (`numIters`) and conversion to microseconds into SampleRunner’s `measure` method.

palimondo committed:
[benchmark] Gardening: Sensibly rename variables

To make sense of this spaghetti code, let’s first use reasonable variable names:
* scale -> numIters
* elapsed_time -> time

palimondo committed:
[benchmark] Gardening: Documentation of numIters

Clarified the need for capping `numIters` according to the discussion at https://github.com/apple/swift/pull/17268#issuecomment-404831035

The sampling loop is a hairy piece of code, because it’s trying to reuse the calibration measurement as a regular sample, in case the computed `numIters` turns out to be 1. But it conflicts with the case when `fixedNumIters` is 1, necessitating a separate measurement in the else branch… That was a quick fix back then, but its hard to make it clean. More thinking is required…

palimondo committed:
[benchmark] Refactor sampling loop with addSample

Extracted sample saving to inner func `addSample`.
Used it to save the `oneIter` sample from `numIters` calibration when it comes out as 1 and continue the for loop to next sample.

This simplified following code that can now always measure the sample with `numIters` and save it.

palimondo committed:
[benchmark] Refactor numIters computation

The spaghetti if-else code was untangled into nested function that computes `iterationsPerSampleTime` and a single constant `numIters` expression that takes care of the overflow capping as well as the choice between fixed and computed `numIters` value.

The `numIters` is now computed and logged only once per benchmark measurement instead of on every sample.

The sampling loop is now just a single line. Hurrah!

Modified test to verify that the `LogParser` maintains `num-iters` derived from the `Measuring with scale` message across samples.

jckarter committed:
SIL: Don't include convenience initializers in vtables.

They should never be dynamically dispatched (unless `required`), so this entry should never be used.
We were accidentally dynamically dispatching to them in convenience-to-convenience `self.init`
delegations; fix that.

palimondo committed:
[benchmark] Extract yield & add resetMeasurements

palimondo committed:
[benchmark] Refactor run runBenchmarks logVerbose

Extracted nested func logVerbose as instance method on SampleRunner.

Internalized the free functions `runBech` and `runBenchmarks` into SampleRunner as methods `run` and `runBenchmarks`.

palimondo committed:
[benchmark] Gardening: Fixed method indentation

palimondo committed:
[benchmark] Rename SampleRunner -> TestRunner

It is now running all the benchmarks, so it’s a TestRunner.

palimondo committed:
[benchmark] Gardening: Even nicer microseconds

palimondo committed:
[benchmark] Report quantiles from samples

The default benchmark result reports statistics of a normal distribution — mean and standard deviation. Unfortunately the samples from our benchmarks are *not normally distributed*. To get a better picture of the underlying probability distribution, this adds support for reporting quantiles.

See https://en.wikipedia.org/wiki/Quantile

This gives better subsample of the measurements in the summary, without need to resort to the use of a full verbose mode, which might be unnecessarily slow.

palimondo committed:
[benchmark] Gardening maxRSS as Int?

palimondo committed:
[benchmark] Option: delta encoded quantiles format

Added `--delta` argument to print the quantiles in delta encoded format, that ommits 0s.

This results in machine and human readable output that highlights modes and is easily digestible, giving you the feel for the underlying probability distribution of the samples in the reported results:

````
$ ./Benchmark_O --num-iters=1 --num-samples=20 --quantile=20 --delta 170 171 184 185 198 199 418 419 432 433 619 620
#,TEST,SAMPLES,MIN(μs),𝚫V1,𝚫V2,𝚫V3,𝚫V4,𝚫V5,𝚫V6,𝚫V7,𝚫V8,𝚫V9,𝚫VA,𝚫VB,𝚫VC,𝚫VD,𝚫VE,𝚫VF,𝚫VG,𝚫VH,𝚫VI,𝚫VJ,𝚫MAX
170,DropFirstArray,20,171,,,,,,,,,,,,,,,,,,,2,29
171,DropFirstArrayLazy,20,168,,,,,,,,,,,,,,,,,,,,8
184,DropLastArray,20,55,,,,,,,,,,,,,,,,,,,,26
185,DropLastArrayLazy,20,65,,,,,,,,,,,,,,,,,,,1,90
198,DropWhileArray,20,214,1,,,,,,,,,,,,,,,,,1,27,2
199,DropWhileArrayLazy,20,464,,,,1,,,,,,,,1,1,1,4,9,1,9,113,2903
418,PrefixArray,20,132,,,,,,,,,,,,,,,,,1,1,32,394
419,PrefixArrayLazy,20,168,,,,,,,,,,,,1,,2,9,1,15,8,88,3338
432,PrefixWhileArray,20,252,1,,,,1,,,,,,,,,,,1,,,,30
433,PrefixWhileArrayLazy,20,168,,,,,,,,,,,,,1,,6,6,14,43,28,10200
619,SuffixArray,20,68,,,,,,,,,,,,,1,,,,22,1,1,4
620,SuffixArrayLazy,20,65,,,,,,,,,,,,,,,,,,1,9,340
````

xedin committed:
[CSSolver] Add skeleton of iterative `solve`

The idea so to split solving into non-recursive steps,
represented by `SolverStep`, each of the steps is resposible
for a unit of work e.g. attempting type variable or
disjunction bindings/choices.

Each step could produce more work via "follow-up" steps,
complete "partial" solution when it's done, or error which
terminates solver loop.

xedin committed:
[CSStep] Implement `mergePartialSolutions`

xedin committed:
[CSStep] Implement `computeFollowupSteps`

palimondo committed:
[benchmark] Doctor explicitly measures memory

Small fix following the last refactorig of MAX_RSS, the `--memory` option is required to measure memory in `--verbose` mode. Added integration test for `check` command of Benchmark_Driver that depended on it.

xedin committed:
[ConstraintSystem] Add `Splitter`, `Component` and binding steps

* `SplitterStep` is responsible for running connected components
  algorithm to determine how many independent sub-systems there are.
  Once that's done it would create one `ComponentStep` per such
  sub-system, and move to try to solve each and then merge partial
  solutions produced by components into complete solution(s).

* `ComponentStep` represents a set of type variables and related
  constraints which could be solved independently. It's further
  simplified into "binding" steps which attempt type variable and
  disjunction choices.

* "Binding" steps such as `TypeVariableStep` and `DisjunctionStep`
  are responsible for trying type binding choices in attempt to
  simplify the system and produce a solution. After attempting each
  choice they introduce a new "split" step to compute more work.

xedin committed:
[CSStep] Add `StepResult` container

nkcsgexi committed:
parser: on demand parsing of class decl members.

nkcsgexi committed:
parser: on demand parsing of struct decl members.

nkcsgexi committed:
parser: on demand parsing of protocol decl members.

xedin committed:
[CSStep] Rename `advance` to `take(bool)` and propagate failures

Failure propagation is crucial for splitter and component steps
because that's the only signal for them to figure out if they
could be completely solved or have to fail.

For example, if one of the component steps created by "split"
fails it would cascade to the rest of the pending component steps
receiving "fail" signal and ultimately result in failure of the
"split" when it's re-taken.

xedin committed:
[CSStep] Add `Disjunction` step

`DisjunctionStep` attempts all of the viable choices,
at least one of the choices has to succeed for disjunction
to be considered a success.

`DisjunctionStep` attempts each choice and generates followup
`SplitterStep` to see if applying choice to constraint system
resulted in any simplification.

xedin committed:
[CSStep] Add License header to `CSStep.cpp`

xedin committed:
[CSStep] Implement `TypeVariable` step

Similar to `DisjunctionStep` `TypeVariableStep` attempts
its choices one-by-one, and if any of the choices match
it makes type variable as a whole successfully solved.

Since type variable attempts choices it means that it
can only produce "split" steps as a follow-up.

xedin committed:
[CSStep] Don't allocate scope if there is only one component

If "split" step produced a single component, there is no reason
to allocate a scope for it and modify constraint system,
because all of the existing type variables and constraints
are related to it anyway, so scope would just end up doing
useless work.

xedin committed:
[CSStep] Split `SolverStep::take(bool)` into `setup`, `take` and `resume`

Introduce a notion of `StepState` with some valid transitions,
and split monolithic `take(bool)` into three phases - `setup`, `take`, and
`resume`.

* `setup` - is preliminary state before the step has been taken
            for the first time.

* `take` - represents actual logic of the step, results in either
           step being done or suspended.

* `resume` - `take` might split steps into smaller ones to make
             incremental progress towards the target, which means
             that the main step has to be suspended. `resume` is
             triggered when all of the "follow-up" steps have been
             executed and are "done".

xedin committed:
[CSStep] NFC: Add some more documentation to `Splitter` and `Component` steps

xedin committed:
[CSStep] Rework how splitter handles component steps

First of all, let the splitter transfer constraints from inactive
work-list to associated components. Also start tracking components
to reinstate constraints back to constraint system when everything
is done, this allows to execute component steps in any desirable order.

xedin committed:
[CSStep] Fix a bug in `DisjunctionStep::resume` caused by was discarding active choice too early

xedin committed:
[CSStep] Add `print(llvm::raw_ostream)` method to `SolverStep`

xedin committed:
[CSStep] Make sure that components are sorter before execution

Execute components with fewer disjunction constraints first,
to make sure that if one of the components fails we do the
smallest amount of work possible.

palimondo committed:
[benchmark] Round quantile idx to nearest or even

Explicitly use round-half-to-even rounding algorithm to match the behavior of numpy's quantile(interpolation='nearest') and quantile estimate type R-3, SAS-2. See:
https://en.wikipedia.org/wiki/Quantile#Estimating_quantiles_from_a_sample

xedin committed:
[CSStep] Restore original best score after component is done

xedin committed:
[CSStep] If previous step before split failed don't do anything

jckarter committed:
SILOptimizer: Turn type(of: self) from uninitialized self stack space into the self argument.

Before we changed convenience inits into allocating entry points, we allowed type(of: self) to be invoked on the uninitialized object, which was fine. This no longer Just Works when self doesn't even exist before `self.init` is called, but to maintain the old semantics and source compatibility, we can still just use the metatype value we were passed in.

xedin committed:
[DisjunctionStep] Start remembering choices when requested

Extract choice attempting logic into `DisjunctionStep::attemptChoice`
and start recording taken choices in constraint system when
requested by disjunction.

xedin committed:
[CSStep] Include related and bound type variables as "in scope"

Re-introduce all type variables which either belong to the given
component or are already bound.

xedin committed:
[CSStep] Rework how orphaned constraints are handled by splitter/component steps

Orphaned constraints can be re-introduced to the system by `setup`
of the component, and all of them could be returned back by
`SplitterStep::resume` there so no need to use destructor for that.

Also, orphan constraints have to be registered as regular constraints
by components as well as marked as "orphaned".

Commit:
[ConstraintSystem] Infer empty closures as returning () more eagerly.

We previously allowed these closures to default to (), but be inferred
as other types as well, which means that we will find some expressions
to be ambiguous because we end up finding multiple viable solutions
where there is really only one reasonable solution.

Fixes: rdar://problem/42337247

compnerd committed:
stdlib: adjust the shims for MSVC mode compilation

The SwiftStdint.h header is used in the compiler as well.  The compiler may be
built with cl (Visual Studio) on Windows, which does not define
`__INTPTR_TYPE__` nor does it define `__INTPTR_WIDTH__`.  Simply define the
`__swift_{,u}intptr_t` typedefs as Microsoft does on that platform when building
with Visual Studio.

jckarter committed:
Reflection: Update extra inhabitants logic for structs.

Structs now use extra inhabitants from whichever field has the most, not only the first field. rdar://problem/44239246

compnerd committed:
Frontend: make `takeSILModule` out-of-line

The use of `std::move` forces the complete definition of the `SILModule` type.
Move the definition out-of-line to allow a forward declaration of `SILModule`
instead.

xedin committed:
[CSStep] Try to simplify constraint system before performing split

compnerd committed:
Demangling: use `assert` over `llvm_unreachable`

Avoid a dependency on LLVMSupport at runtime through the `llvm_unreachable`.
This would pull in `llvm_unreachable_internal` in debug builds, which requires a
runtime dependency on LLVMSupport which increases the size of the binary
considerably.

xedin committed:
[CSStep] Filter solutions only if component is part of a large split

nkcsgexi committed:
Parser: disable lexer diagnostics during delayed parsing.

The lexer diagnostics have already been emitted during token skipping.

nkcsgexi committed:
Fix an unused variable warning in no-assertion build.

DougGregor committed:
[ABI] Emit associated type descriptors referencing each requirement.

Emit associated type descriptors (as aliases) to reference each associated
type requirement within a resilient protocol.

compnerd committed:
AST: friend SimpleRequest without type specifier

When building with Visual Studio, the following error is identified:

  'SimpleRequest': non-class template has already been declared as a class template
  note: see declaration of 'SimpleRequest'

Change the friend declarations to friend the type without the `class` specifier.

nkcsgexi committed:
Parser: disable delayed member decl parsing when parsing SIL.

nkcsgexi committed:
swift-ide-test: input complete test is for REPL.

jrose-apple committed:
[CMake] Remove unnecessary guard for standalone overlay builds

These should install the swiftmodule and swiftdoc files too.

jrose-apple committed:
[CMake] Install swiftinterface files alongside swiftmodules

This may not be their final destination, but at least this will
get them into built toolchains, which means it'll be easier to
do integration testing once all the other pieces are in place.

rdar://problem/43823937

Commit:
Applied Jordan’s patches to tests.

jrose-apple committed:
[Serialization] Compact the Offset value in PartiallySerialized

serialization::BitOffset is set up for being in a PointerUnion, but in
this case that's not just overkill but actually wasteful, since we
have an extra flag to pack in. Use a raw bitfield instead.

No functionality change.

DougGregor committed:
[ABI] Add protocol requirements base descriptor.

Introduce an alias that refers one element prior to the start of a
protocol descriptor’s protocol requirements. This can be subtracted from
an associated type descriptor address to determine the offset of the
associated type accessor within a corresponding witness table. The code
generation for the latter is not yet implemented.

slavapestov committed:
SILGen: VTable and witness thunk emission uses new function type representation

This eliminates the only usage of AbstractionPattern::dropLastTupleElement().

For now, we keep the argument tuple-based entry points in TransformArguments
as well, because they're used by re-abstraction thunks. Those require a bit more
refactoring.

AnthonyLatsis committed:
[Sema] Suppress deprecation warnings in synthesized code

jrose-apple committed:
[Serialization] getIdentifierText, for when an Identifier isn't needed

Not everything that goes into the "identifier table" is actually an
Identifier. If we don't need the string to be uniqued in the
ASTContext, don't bother calling getIdentifier.

May save some memory usage (and a string table lookup) for
deserialization.

DougGregor committed:
[IRGen] witness_method should always refer to a slot in the witness table.

compnerd committed:
AST, SIL: use `llvm::StringSwitch` over cascading if/else if

Visual Studio's compiler has a hard limit on the number of if/else if cascading
blocks at 128 elements (https://msdn.microsoft.com/en-us/library/dcda4f64.aspx).
Use `llvm::StringSwitch` which is both easier to read (although less imported in
this case due to the macro generation) and allows us to avoid this limit.

compnerd committed:
tapi: make it build on Windows

Since TAPI is currently built statically, just define away `TAPI_PUBLIC` on
Windows.  If tapi is to be built dynamic, we would need to annotate dllstorage
on the public interfaces.

akyrtzi committed:
[test/lit.cfg] Use absolute imports

This fixes an issue when using Docker on OSX, 'import shutil' in 'lit.cfg' ends up finding lit.ShUtil instead of the module from the standard library.
Switching to absolute imports ensures that the module from the standard library will be used.

akyrtzi committed:
[test/Index/Store] For a test, replace 'XFAIL: linux' with 'REQUIRES: objc_interop'

compnerd committed:
SILOptimizer: use `LLVM_ATTRIBUTE_USED`

This makes the use of the used attribute more portable.
`__attribute__((__used__))` is not accepted by Visual Studio, but LLVM
conveniently provides the helpful macro `LLVM_ATTRIBUTE_USED` to enable this.

DougGregor committed:
[IRGen] Implement resilient associated type witness access pattern.

When accessing the associated type witness metadata for a resilient
protocol, compute the index based on the difference between the
associated type’s descriptor and the protocol requirement base descriptor
to determine the offset into the witness table.

compnerd committed:
Sema: qualify the `flags` member

When building with Visual Studio, the flags member would not be found.  Qualify
the member with the scope (class name) to resolve the ambiguity.

compnerd committed:
build: improve the builds with Visual Studio

Visual Studio's compiler does not accept `-Werror=switch`.  Use the equivalent
`-we4062` option instead.  Avoid using the `--` separate that clang uses to
identify that the options that follow are files and not options on the Visual
Studio compiler to reduce the unnecessary spew when building on Windows.

mikeash committed:
[Runtime][Stdlib] Use dedicated thread-local storage keys when possible.

rdar://problem/32275323 rdar://problem/44104305

Commit:
[AST] Preserve sugar for generic typealiases.

<rdar://problem/43110802>

dcci commented:
@swift-ci test and merge

jrose-apple commented:
Isn't this not correct because we've lost the parent type? Or am I misremembering? @slavapestov

nkcsgexi committed:
ASTVerifier: avoid verifying IterableDeclContext if it has unparsed members.

nkcsgexi committed:
scale-test: teach the test driver to test parse only.

jrose-apple committed:
[Serialization] Store all SIL names in the identifier table

We were already doing this for cross-references in SIL function bodies,
but by doing it for the SIL index tables too we can shrink the size of
the module file a bit. (Not that this is a super-important metric, but
still.)

nkcsgexi committed:
test: add a scale test for the number of IterableDeclContext getting parsed.

jrose-apple committed:
[CMake] Stop compiling API notes files to a binary format (#19205)

Clang's been able to read the textual format for a long time, and
indeed that's what's used in the SDK. This isn't even really a
performance win because the information is cached in PCMs.

rdar://problem/34293901

Commit:
Fixed test to reflect specificity of “could not build Objective-C module” error.

dcci commented:
@swift-ci please clean test

jrose-apple committed:
[Serialization] Fix caching of deserialized Identifiers (#19264)

This was causing us to do a string table lookup on every use of a
serialized identifier, even a repeated one.

jckarter committed:
Update tests based on review feedback.

dcci commented:
@swift-ci please clean test macos

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -3022,14 +3022,13 @@ void TypeAliasDecl::setUnderlyingType(Type underlying) {
+        genericSig, [&](SubstitutableType *type) -> Type { return type; },
+        MakeAbstractConformanceForGenericType());
+
+    auto sugaredType = NameAliasType::get(this, Type(), subs, underlying);
```
The parent type should probably be the declared interface type of the parent DC, if the parent is a type context

nkcsgexi committed:
FrontendTool: parse all delayed member decls when we're dumping parse tree or verifying parse diagnostics.

jrose-apple committed:
[Serialization] Add a separate string table for non-Identifiers

They're not actually stored separately in the module file, nor
deserialized separately, but this way we're not sticking a bunch of
strings in the /ASTContext's/ string table, which would persist after
serialization. (The ASTContext's string table is also implemented
using a BumpPtrAllocator-backed StringMap, so the performance
characteristics of this should be about the same.)

DougGregor committed:
[Runtime] Fix _ObjectiveCBridgeableWitnessTable::ObjectiveCType

The signature of this function pointer was incorrect, because it had not
yet been updated for the metadata request/response architecture. We
were getting lucky (!).

jrose-apple committed:
[Serialization] Remove the last use of Identifiers for SIL names

The string-keyed tables don't actually need Identifier keys on the
serialization side -- no reason to persist these in the ASTContext's
string table either.

xedin committed:
[CSStep] Record all of the type vars related to component while creating a split

Before this change some of the type variables, especially the ones
that have been previously bound haven't been associated with any
of the component steps, so it had to be done by scope once per step,
but now such recording is done once per split.

DougGregor committed:
[ABI] Emit associated type witnesses resiliently.

Emit associated type witnesses into resilient conformance tables, so they
can be re-ordered within the protocol without breaking clients. This will
(eventually) permit adding new, defaulted associated types to protocols
resiliently.

xedin committed:
[CSStep] Use # of disjunctions directly to determine component ordering

slavapestov committed:
SIL: Don't implode parameters into a tuple in DestructureInputs::visitTopLevelParams()

When lowering an AST function type to a SIL function type we would
implode the parameter list into a tuple, then visit the tuple, which
would explode it back again unless we had an opaque abstraction
pattern.

Instead, only implode the parameter list if we have an opaque
abstraction pattern, otherwise visit each parameter directly.

A few related cleanups:

- Instead of calling AnyFunctionType::Param::getType() which returns
  an InOutType, look at ParameterFlags directly, avoiding InOutType
  altogether.

- Use AbstractionPattern::getFunctionParamType() instead of
  getFunctionInputType().getTupleElementType(), which allows us to
  distinguish multiple parameters from a sigle tuple parameter.
  This distinction does not matter yet but it will soon.

xedin committed:
[ConstraintSystem] Rename `solveIteratively` to `solve` and start using it

Remove solution filtering from new iterative `solve` method, and
replace all usages of `solveRec` with it.

xedin committed:
[ConstraintSystem] Remove recursive solver algorithm and related logic

`solveRec` is now gone as well as all related logic,
such as attempting type variable and disjunction choices recursively.

slavapestov committed:
SIL: Don't implode parameters into a tuple in TypeConverter::getLoweredFormalTypes()

getLoweredFormalTypes() behaved differently for top-level
functions and methods:

- For functions it would bridge each parameter type and return a new
  function type with the bridged parameters.

- For methods, it would build a new function type whose argument list
  was always a tuple of the form ((Args...), Self), with all formal
  parameters of the method imploded into a single tuple argument.

This is a simple change in isolation, but AbstractionPatterns for
Objective-C methods and C functions imported as methods depended on
the old representation.

This patch reworks AbstractionPattern:getFunctionParamType() to do
the right thing for these AbstractionPattern kinds.

Note that I haven't touched the older getFunctionInputType() which
still does the old thing, but it's going away.

slavapestov committed:
SIL: Implement remaining cases for AbstractionPattern::getFunctionParamType()

slavapestov committed:
SIL: Remove AbstractionPattern::dropLastTupleElement()

AnthonyLatsis committed:
Add test for usage of deprecated inherited init

jrose-apple committed:
[Serialization] Use local BumpPtrAllocator for fixed-sized buffers

...instead of std::vector, which (1) will always make separate
allocations, and (2) has features and overhead we don't need

I don't expect this to actually affect performance too much, but it
seems more correct for what Serialization needs anyway.

benlangmuir committed:
Disable api-digester/diagnostics.swift, which has been failing in CI

benrimmington committed:
[SE-0211] Emoji properties require ICU 57 or later

khanlou committed:
Add `count(where:)` and tests (#16099)

* add count(where:) and tests

* Revise count(where:) documentation

* Remove errant word in abstract

* add a benchmark for ranges and strings with help from @natecook1000

* update benchmark to use Array instead of Range

jrose-apple committed:
[Driver] Remove handrolled StringSetIterator in favor of 'keys()'

An llvm::StringSet is "just" an llvm::StringMap where the values are
ignored, so the 'keys()' range already does what we need.

No functionality change.

jrose-apple committed:
Use llvm::StringSet instead of StringMap where appropriate

That is, where we aren't using the value for anything.

No functionality change.

jrose-apple committed:
Replace StringMap with DenseMap when the keys don't need to be owned

StringMap always copies its strings into its own storage. A DenseMap
of StringRefs has the same caveats as any other use of StringRef, but
in the cases I've changed the string has very clear ownership that
outlives the map.

No functionality change, but should reduce memory usage and malloc
traffic a little.

DougGregor committed:
[IRGen] Simplify nondependent associated type witness metadata accessors.

When we’re creating an associated type witness metadata accessor for
resilience reasons, but the associated type witness doesn’t involve any
type parameters, directly form the type metadata reference (and don’t
cache it).

While here… update all of the IRGen/SILGen test cases perturbed by the
introduction of resilient associated type access patterns.

palimondo committed:
[benchmarks] Rename column label SPEEDUP to RATIO

Since the results comparisons are now used to also compare code sizes in addition to runtimes, it makes sense to rename the column label to the more neutral term “ratio” instead of old “speedup”.

slavapestov committed:
SIL: Simplify AbstractionPattern::getFunctionParamType()

Use the new AnyFunctionType::Param::getParameterType() method
John added, instead of encoding knowledge of varargs directly
here.

Also don't wrap the result in an InOutType since for now,
clients only use this for Objective-C methods which don't
have inout parameters, and upcoming usages need the unwrapped
type anyway.

DougGregor committed:
[ABI] Use associated type descriptors for generic parameter references.

Generic parameter references, which occur in generic requirement
metadata, were hardcoding associated type indices. Instead, use
relative references to associated type descriptors and perform the
index calculation at runtime.

Associated types can now be reordered resiliently (without relying on 
sorting), which is the first main step toward rdar://problem/44167982.

compnerd committed:
litter the tree with `llvm_unreachable`

This silences the instances of the warning from Visual Studio about not all
codepaths returning a value.  This makes the output more readable and less
likely to lose useful warnings.  NFC.

jckarter committed:
Give opaque existential containers extra inhabitants.

We can use the extra inhabitants of the type metadata field as extra inhabitants of the entire
existential container, allowing `Any?` and similar types to be the same size as non-optional
existentials.

compnerd committed:
IRGen: use the `llvm::TrailingObjects` base class explicitly

Visual Studio is unable to follow the inheritance to find the method and objects
with:

  lib\IRGen\GenExistential.cpp(153): error C2903: 'totalSizeToAlloc': symbol is neither a class template nor a function template
  lib\IRGen\GenExistential.cpp(153): note: This diagnostic occurred in the compiler generated function 'const Derived *`anonymous-namespace'::ExistentialTypeInfoBase<Derived,Base>::create(llvm::ArrayRef<T>,As &&...)'
        with
        [
            T=const swift::ProtocolDecl *
        ]
  lib\IRGen\GenExistential.cpp(202): note: see reference to class template instantiation '`anonymous-namespace'::ExistentialTypeInfoBase<Derived,Base>' being compiled
  lib\IRGen\GenExistential.cpp(154): error C2760: syntax error: unexpected token 'const', expected 'expression'
  lib\IRGen\GenExistential.cpp(154): note: This diagnostic occurred in the compiler generated function 'const Derived *`anonymous-namespace'::ExistentialTypeInfoBase<Derived,Base>::create(llvm::ArrayRef<T>,As &&...)'
        with
        [
            T=const swift::ProtocolDecl *
        ]

jckarter committed:
Reserve the spare bits in closure context pointers.

Although this is a refcounted pointer (for escaping closures, at least), we'd like to eventually take advantage of the ability to bit pack arbitrary payloads into refcountable fields on 64-bit platforms. For nonescaping closures, the context ought to be a trivial arbitrary word as well, so we shouldn't be looking for spare bits to begin with.

compnerd committed:
SILOptimizer: make a conversion operation explicit

Visual Studio objects to the existing construct with:

  swift/SILOptimizer/Analysis/ARCAnalysis.h(233): warning C4927: illegal conversion; more than one user-defined conversion has been implicitly applied
  swift/SILOptimizer/Analysis/ARCAnalysis.h(233): note: while calling the constructor 'llvm::Optional<llvm::ArrayRef<swift::SILInstruction *>>::Optiona(T &&)'
	with
	[
	  T=llvm::ArrayRef<swift::SILInstruction *>
	]
  llvm/ADT/Optional.h(143): note: see declaration of 'llvm::Optional<llvm::ArrayRef<swift::SILInstruction *>>::Optional'

jckarter committed:
Small cleanups to type(of: self) handling in DI

Should be NFC

DougGregor committed:
[AST] Kill rarely-used DenseMap from ProtocolDecl.

Put a DenseMap onto the ASTContext’s implementation instead.

graydon committed:
[Stats] Add SWIFTC_MAXIMUM_DETERMINISM to inhibit parallelism everywhere.

compnerd committed:
Sema: quote mark name (NFC)

Visual Studio improperly tries to parse the unknown pragma.  Quote the parameter
to allow building.

  lib\Sema\TypeCheckDeclObjC.cpp(970): error C2018: unknown character '0x40'

xedin committed:
[CSStep/TypeVar] Check if loop should be stopped even if current binding has failed

Eagerly check if optimal solution has already been found even if
current binding attempt has failed, because otherwise it might
check too many bindings which would result in ambiguity.

jrose-apple committed:
[test] Fix mixed-target-using-module.swift autolinking checks (#19308)

Negative tests aren't great to begin with, and this one was looking
for a pattern that could no longer happen. This particular example
doesn't seem to have regressed, but something similar has that I'm
still looking into.

johnno1962 committed:
[Parse] Allow multiline attribute messages (SE-200) (#19219)

Multiline string literal at attribute message position was disallowed in
59778f8ecb5f69bd4dda496b07a67769e5d37d74.

Reworked to try to at least get multiline strings working which might be
useful as messages for attributes (for example a detailed “unavailable”
annotation) minus the code which read off the start of the StringRef buffer.

Commit:
Add -driver-force-response-files to enable testing.

slavapestov committed:
SILGen: Remove ArgumentSource::getSILSubst{RValue,}Type()

slavapestov committed:
SILGen: Remove InOutType usages from ArgEmitter

slavapestov committed:
SILGen: Remove ArgumentSource::getSubstType()

This pushes an InOutType usage down to PreparedArguments, which still
needs to be refactored to use AnyFunctionType::Param.

slavapestov committed:
SIL: Refactor a usage of AbstractionPattern::transformType()

slavapestov committed:
SIL: Remove AbstractionPattern::transformType()

Commit:
Fix spaces in filenames in response files.

Commit:
add test

compnerd committed:
AST: SFINAE away template'd constructor

Because we cannot explicitly control which constructor is used, and Visual
Studio also instantiates the templated constructor (even with the copy and move
constructors available, we end up with ambiguity in the constructor when trying
to instantiate a `llvm::SetVector<AnyRequest>` in the Evaluator.

```
	swift\include\swift\AST\AnyRequest.h(98): error C2027: use of undefined type 'swift::TypeID<Request>'
		with
		[
		    Request=ValueType
		]
	swift\include\swift\AST\AnyRequest.h(98): note: see declaration of 'swift::TypeID<Request>'
		with
		[
		    Request=ValueType
		]
	swift\include\swift\AST\AnyRequest.h(97): note: while compiling class template member function 'swift::AnyRequest::Holder<ValueType>::Holder(const Request &)'
		with
		[
		    Request=ValueType
		]
	swift\include\swift\AST\AnyRequest.h(167): note: see reference to function template instantiation 'swift::AnyRequest::Holder<ValueType>::Holder(const Request &)' being compiled
		with
		[
		    Request=ValueType
		]
	swift\include\swift\AST\AnyRequest.h(168): note: see reference to class template instantiation 'swift::AnyRequest::Holder<ValueType>' being compiled
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\xmemory0(917): note: see reference to function template instantiation 'swift::AnyRequest::AnyRequest<T&>(swift::AnyRequest&)' being compiled
		with
		[
		    T=swift::AnyRequest
		]
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\xmemory(120): note: see reference to function template instantiation 'void std::_Default_allocator_traits<_Alloc>::construct<_Ty,T&>(_Alloc &,_Objty *const ,T &)' being compiled
		with
		[
		    _Alloc=std::allocator<swift::AnyRequest>,
		    _Ty=swift::AnyRequest,
		    T=swift::AnyRequest,
		    _Objty=swift::AnyRequest
		]
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\xmemory(120): note: see reference to function template instantiation 'void std::_Default_allocator_traits<_Alloc>::construct<_Ty,T&>(_Alloc &,_Objty *const ,T &)' being compiled
		with
		[
		    _Alloc=std::allocator<swift::AnyRequest>,
		    _Ty=swift::AnyRequest,
		    T=swift::AnyRequest,
		    _Objty=swift::AnyRequest
		]
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\xmemory(153): note: see reference to function template instantiation '_FwdIt std::_Uninitialized_copy_al_unchecked<_Iter,_Iter,_Alloc>(_InIt,_InIt,_FwdIt,_Alloc &,std::_General_ptr_iterator_tag,std::_Any_tag)' being compiled
		with
		[
		    _FwdIt=swift::AnyRequest *,
		    _Iter=swift::AnyRequest *,
		    _Alloc=std::allocator<swift::AnyRequest>,
		    _InIt=swift::AnyRequest *
		]
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\vector(1939): note: see reference to function template instantiation '_FwdIt *std::_Uninitialized_copy<swift::AnyRequest*,swift::AnyRequest*,std::allocator<_Ty>>(_InIt,_InIt,_FwdIt,_Alloc &)' being compiled
		with
		[
		    _FwdIt=swift::AnyRequest *,
		    _Ty=swift::AnyRequest,
		    _InIt=swift::AnyRequest *,
		    _Alloc=std::allocator<swift::AnyRequest>
		]
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\vector(1938): note: while compiling class template member function 'void std::vector<T,std::allocator<_Ty>>::_Umove_if_noexcept1(swift::AnyRequest *,swift::AnyRequest *,swift::AnyRequest *,std::false_type)'
		with
		[
		    T=swift::AnyRequest,
		    _Ty=swift::AnyRequest
		]
	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.13.26128\include\vector(1944): note: see reference to function template instantiation 'void std::vector<T,std::allocator<_Ty>>::_Umove_if_noexcept1(swift::AnyRequest *,swift::AnyRequest *,swift::AnyRequest *,std::false_type)' being compiled
		with
		[
		    T=swift::AnyRequest,
		    _Ty=swift::AnyRequest
		]
	llvm\include\llvm\ADT\SetVector.h(49): note: see reference to class template instantiation 'std::vector<T,std::allocator<_Ty>>' being compiled
		with
		[
		    T=swift::AnyRequest,
		    _Ty=swift::AnyRequest
		]
	swift\include\swift\AST\Evaluator.h(208): note: see reference to class template instantiation 'llvm::SetVector<swift::AnyRequest,std::vector<T,std::allocator<_Ty>>,llvm::DenseSet<T,llvm::DenseMapInfo<swift::AnyRequest>>>' being compiled
		with
		[
		    T=swift::AnyRequest,
		    _Ty=swift::AnyRequest
		]
	swift\include\swift\AST\AnyRequest.h(97): error C2065: 'value': undeclared identifier
```

Hoist the type alias into the template specialization so that we can use that
in the SFINAE expression rather than duplicating it.

Commit:
remove extra newlines at end.

Commit:
Use %target-build-swift and -parse.

harlanhaskins committed:
[InterfaceGen] Print bodies of inlinable functions in textual interfaces (#19224)

* Introduce stored inlinable function bodies

* Remove serialization changes

* [InterfaceGen] Print inlinable function bodies

* Clean up a little bit and add test

* Undo changes to InlinableText

* Add serialization and deserialization for inlinable body text

* Allow parser to parse accessor bodies in interfaces

* Fix some tests

* Fix remaining tests

* Add tests for usableFromInline decls

* Add comments

* Clean up function body printing throughout

* Add tests for subscripts

* Remove comment about subscript inlinable text

* Address some comments

* Handle lack of @objc on Linux

DougGregor committed:
[SIL] Unify default witness table entries with witness table entries.

SILWitnessTable::Entry already contains a superset of what was supported
by SILDefaultWitnessTable::Entry, the latter of which only had “no entry”
and “method” states. Make SILDefaultWitnessTable::Entry an alias for
SILWitnessTable::Entry, and unify all of the parsing/printing/
(de)serialization logic.

xedin committed:
[ConstraintSystem] Remove unnecessary forward declarations of the steps

Also move `DisjunctionStep` from being a friend of `ConstraintSystem`
because it's not strictly necessary.

compnerd committed:
stdlib: repair the Windows runtime build

Fix syntactic issues in the swift runtime build from the recent TLS changes.
Move the helper definitions into the TLS stubs rather than emitting them
everywhere.  This allows the runtime to build again on Windows.

xedin committed:
[ConstraintSystem] Integrate emergency break into iterative solver loop

There are situations when solver goes exponential, before
(when solver was recursive) it would fail relatively quickly by running
out of stack space, now we really need to make sure that emergency
break is in place in main solver loop, otherwise we risk it running for
a really long time if doesn't consume too much memory since default
timeout is ~10 mins.

slavapestov committed:
SILGen: Rename ArgumentSource::requiresCalleeToEvaluate() to ArgumentSource::isShuffle()

Also, move the check to make it explicit that only a TupleShuffleExpr
at the top level goes through the argument emission code path; a
TupleShuffleExpr appearing inside a ParenExpr or TupleExpr is a
tuple conversion, which is totally unrelated and emitted as an RValue.

This is unfortunate and we should split off ArgumentShuffleExpr from
TupleShuffleExpr, and eventually, fold ArgumentShuffleExpr into
ApplyExpr.

slavapestov committed:
SILGen: Remove dead code from Swift 3 removal

benlangmuir committed:
Revert "[benchmark] Report Quantiles from Benchmark_O and a TON of Gardening"

jrose-apple committed:
[ModuleInterface] Don't include 'import Builtin' for the stdlib (#19319)

The special 'Builtin' module is implicitly imported via -parse-stdlib,
and can't be found via normal import resolution, so we shouldn't print
it in a textual interface.

slavapestov committed:
Serialization: Serialize both the substituted and unsubstituted type for a NameAliasType

slavapestov committed:
Add scale test for nested type extension binding

This performance problem was fixed a while ago by Doug's request
evaluator work.

jrose-apple committed:
[docs] Add "WIP" and "DNM" to Lexicon.rst (#19329)

"WIP" is hardly unique to Swift, but it is a three-letter acronym that
someone might not have encountered before. "DNM" seems to have arisen
organically over the last few weeks.

adrian-prantl committed:
Zero-initialize -Onone debug shadow copies for exploded values
and simplify the initialization code by emitting a memset intrinsic.

rdar://problem/43566087

DougGregor committed:
[ABI] Add default associated type witnesses to resilient protocols.

When an associated type witness has a default, record that as part of
the protocol and emit a default associated type metadata accessor into the
default witness table. This allows a defaulted associated type to be
added to a protocol resiliently.

This is another part of rdar://problem/44167982, but it’s still very
limiting because the new associated type cannot have any conformances.

DougGregor committed:
Adjust test for Linux

DougGregor committed:
[IRGen] Fix associated type resilience for 32-bit platforms.

xedin committed:
[CSStep] Add `isDebugMode` and `getDebugLogger` methods and refactor logging

Instead of using `TypeChecker` and related APIs directly every time,
let's just abstract it all away to easily check if solver is running
in debug mode and to produce indented logger.

flashspys committed:
Fixed typo in example code

DougGregor committed:
Another test fix for Linux

slavapestov committed:
SILGen: Remove ArgumentSource::Kind::Tuple

slavapestov committed:
SILGen: Change PreparedArguments to use an array of AnyFunctionType::Param

This solves a recently-added TODO, eliminating two usages of
FunctionType::getInput() from SILGen.

slavapestov committed:
SILGen: Remove ArgumentSource::rewriteType()

slavapestov committed:
SILGen: Move usages of AbstractionPattern::getFunctionInputType() closer to their uses

DougGregor committed:
[SIL Parser] Handle parsing of ‘Self’ in default witness tables.

There is no good way to do this, so take a shortcut for now of 
pre-resolving ‘Self’.

DougGregor committed:
Resolve two crashers and fix a test issue on Linux.

compnerd committed:
AST: remove extra copy constructors

Visual Studio objects with the following:

```
  AST/AnyRequest.h(243): error: C2580: 'swift::AnyRequest::AnyRequest(const swift::AnyRequest &)': multiple versions of a defaulted special member functions are not allowed
  AST/AnyRequest(243): warning C4521: 'swift::AnyRequest': multiple copy constructors specified
```

Remove the non-const implementation, the default'ed definition should be
sufficient.  Additionally, remove the nonsense const move constructor.

Commit:
[AST] Don't lose the parent type in setUnderlyingType.
